In [1]:
#import libraries
import numpy as np
import pandas as pd
import scipy.stats as stats

## Month 2 - exams
## Question 1 - Pandas for Data Analysis

In [2]:
#Import csv file
df = pd.read_csv('https://raw.githubusercontent.com/ek-chris/Practice_datasets/refs/heads/main/eletronic_sales.csv')
df['Date'] = pd.to_datetime(df['Date'])
df['Year'] = df['Date'].dt.year
df['Month'] = df['Date'].dt.month
df['Day'] = df['Date'].dt.day
df['Day_of_Week'] = df['Date'].dt.day_of_week
df.head()

Date Branch Sales Agent Products  Units   Price  Year  Month  Day  \
0 2014-09-01   Woji     Chinedu    Apple      2  125.00  2014      9    1   
1 2015-06-17   Woji       Emeka    Apple      5  125.00  2015      6   17   
2 2015-09-10   Woji     Ibrahim   Lenovo      7    1.29  2015      9   10   
3 2015-11-17   Woji        Tolu       HP     11    4.99  2015     11   17   
4 2015-10-31   Woji       Tonye   Lenovo     14    1.29  2015     10   31   

   Day_of_Week  
0            0  
1            2  
2            3  
3            1  
4            5

### b. Branch-Level Total Sales

In [3]:
df['Total Sales'] = df['Units'] * df['Price']
df_branch = df.groupby('Branch')['Total Sales'].sum().reset_index()
df_branch[['Branch', 'Total Sales']]

Branch  Total Sales
0    GRA      6002.09
1   Town      2486.72
2   Woji     11139.07

### c. Top Performing Sales Agents

In [4]:
df.groupby('Sales Agent')['Total Sales'].sum().reset_index()

Sales Agent  Total Sales
0     Blessing      2363.04
1      Chinedu      1641.43
2       Chioma      3102.30
3        Emeka      3109.44
4       George      1387.77
5      Ibrahim      1749.87
6         Tolu      2812.19
7        Tonye       438.37
8      Torbari       536.75
9        Tunde      1203.11
10        Uche      1283.61

### d. Introducing and Filling Missing Values

In [ ]:
df.loc[[5, 15, 25], 'Price'] = np.nan
df.fillna({'Price': df['Price'].median()}, inplace = True)
df

### e.  Product-Level Summary

In [10]:
df_agg = df.groupby('Products').agg({'Price': 'mean', 'Units': 'sum'}).rename(columns = {'Price': 'average price', 'Units': 'total units sold'})
df_agg

average price  total units sold
Products                                 
Apple        175.000000                10
Compaq         5.190000               278
Dell          11.912857               395
HP            11.524000               722
Lenovo         3.005385               716

## Question 2 - Numpy for Numeric Computation

### a. Array Creation and Basic Manipulation

In [27]:
nums = np.random.randint(10, 100, 20)
nums_reshape = nums.reshape(4, 5)
print('Reshaped Matrix')
print(nums_reshape)
print('\n')
print('First 2 rows and Last 3 columns')
print(nums_reshape[:2, -3:])
print('\n')
print('Mean:', nums.mean())
print('Standard Deviation:', nums.std())

Reshaped Matrix
[[24 26 83 33 29]
 [10 68 78 90 58]
 [91 29 85 42 36]
 [20 99 56 88 34]]


First 2 rows and Last 3 columns
[[83 33 29]
 [78 90 58]]


Mean: 53.95
Standard Deviation: 28.084648831701635


In [19]:
nums_reshape

array([[99, 76, 59, 90, 93],
       [85, 29, 70, 90, 41],
       [97, 87, 30, 29, 25],
       [93, 36, 24, 25, 51]], dtype=int32)

### b. Operations on 2D Arrays

In [36]:
scores = np.random.randint(30, 100, size = (10, 5))
print('Average Score per student')
print(scores.mean(axis=1))
print('Maximum score')
print(scores.max())
print('Minimum score')
print(scores.min())

Average Score per student
[61.6 76.2 60.8 73.4 77.6 53.  84.4 75.4 55.4 59.2]
Maximum score
98
Minimum score
30


### c. Working with 3D Arrays

In [45]:
ints = np.random.randint(1, 20, size = (3, 4, 2))
print(ints.sum(axis=1))
print(ints.max(axis = 2))
print(ints.flatten())

[[42 36]
 [51 28]
 [47 47]]
[[10 12 17 19]
 [18 19  8 13]
 [16 19  1 16]]
[ 5 10 12  8  6 17 19  1 18  3 12 19  8  5 13  1 12 16 18 19  1  1 16 11]


# Question 3 - Statistics for statistical Analysis

### a. Measures of Center and Spread

#### Mean
CO2 emisions is given by the dataset: D = [25.4, 30.2, 22.5, 28.1, 35].
To compute the mean, we use the formula
$$
\mu = \frac{\sum_{i=0}^n (X_i \in D)}{n}
$$

$$
\implies \mu = \frac{141.2}{5}  = 28.24
$$

$$
\therefore \mu = 28.24
$$
#### Median
Median  is given by the mid value after sorting, since the size of our data is 5, we take the mid value
$$
\text{i.e  } median = 28.1
$$
#### 

#### Mode
There is no mode in our dataset



### b. Hypothesis testing

In [58]:
### b. Hypothesis testing

arg = [60, 62, 58, 63, 59]
bang = [15, 12, 18, 14, 16]

#Define the Null and Alternate Hypthesis
#H0 = There is no significant difference in mean beef consumption 
#H0 = There is significant difference in mean beef consumption 

tstat, pvalue = stats.ttest_ind(arg, bang)
print(f't statistic: {tstat}, Pvalue: {pvalue}')

# Since the p value is less than significant level 0.05, then there is enough statistical evidence to reject the null hypothesis, that is statistical difference in mean beef consumption between both countries


t statistic: 33.288895109847196, Pvalue: 7.237288101315669e-10


### c. Correlation Analysis

In [ ]:
con = [10, 15, 20, 25, 30]
co2 = [30, 45, 50, 70, 85]

t_stat, p_val = stats.pearsonr(con, co2)
print(f'Correlation coefficient: {t_stat}, Pvalue: {p_val}')


#b. The correlation coefficient is 0.98 which means we have a very strong correlation
#c. Corellation of approximately 0 implies there is no correlation between both features

Correlation coefficient: 0.9872177725716262, Pvalue: 0.001731447097541322


## 4. Linear Algebra

### a. Total score per subject

In [76]:
scores = np.array([[80, 70, 90], [60, 85, 75], [95, 88, 92], [70, 60, 65]])
print(scores)
# Total Scores for each student
print('\n')
print('Total Scores for each student')
print(np.sum(scores, axis = 1))
weight = np.sum(scores, axis = 1)
print(weight/3)




[[80 70 90]
 [60 85 75]
 [95 88 92]
 [70 60 65]]


Total Scores for each student
[240 220 275 195]
[80.         73.33333333 91.66666667 65.        ]


### b. Average score per subject

In [69]:
print('Average Scores for each subject')
print(np.sum(scores, axis = 0))

Average Scores for each subject
[305 303 322]


### c. Weighted  Final Score

In [75]:
w = [0.5, 0.3, 0.2]
weighted = scores * w
print('Weighted Average')
print(np.sum(weighted, axis = 1))

Weighted Average
[79.  70.5 92.3 66. ]
